### PR TITLE
RSDK-2639 - Improve fake model for bases

### PR DESCRIPTION
**Pull request by raybjork**
This work adds functionality to the `fake` Base model, bringing it up to par with the corresponding `fake` Arm model.  This is a necessary precursor to being able to write software tests the implementation for `motion.MoveOnMap` as we need a Base model that we use to mock the functions comprising the `InputEnabled` interface

Additionally, the implementation of `kinematicWheeledBase` has been changed slightly.  Functionality used to extract the limits of a slam map have been abstracted away into its own function and moved into the slam package as a helper function.

### CODE REVIEW

File: services/slam/slam.go
Comment (nicksanford) [inline services/slam/slam.go:126]: [q] Is there a ticket for adding this function?

Comment (nicksanford) [reply services/slam/slam.go:126]: [q] why is this only returning the limits in the XZ plane?

Comment (nicksanford) [reply services/slam/slam.go:126]: `[]referenceframe.Limit` doesn't seem very type-safe to me. In order for the caller to get a value they can do something with it needs to specifically be a `[]referenceframe.Limit` of size 2 right?

Is there a reason why we are not using an array of size 2 or a struct that makes it explicit that this is measuring the min & max dimensions in the XZ plane?

Something like:
```
type XYZLimit struct {
X referenceframe.Limit
Y referenceframe.Limit
Z referenceframe.Limit
}
```

**Review by nicksanford — CHANGES_REQUESTED** [review]

File: components/base/fake/fake.go
Comment (randhid) [inline components/base/fake/fake.go:40]: [minor] Since this is actually doing something now, let's use `NewBase` as a constructor since we don't have two points of  truth for the registry and internal to RDK.
```suggestion
				return NewBase(ctx, config, logger)
```

File: components/base/fake/fake.go
Comment (randhid) [inline components/base/fake/fake.go:26]: [major] I'd rather have the default be wheeled base right now without configuration parameters at all exposed to the user.

File: services/slam/slam.go
Comment (nicksanford) [reply services/slam/slam.go:126]: Output of convo:
I think that this function should either exist in the package of the service that uses slam given that this function as written doesn't allow a user to access or perscribe the point cloud who's limits were analyzed. If this function ends up being duplicated in more than one other package I think we can consider adding it to the slam.go.

File: components/base/fake/fake.go
Comment (randhid) [reply components/base/fake/fake.go:26]: It seems to be using `FrameConfig` anyway. We don't need to add those other bases, they're experimental and likely will have to use SetVelocity or SetPower. Let's keep work as narrow as possible.

File: components/base/wheeled/kinematics.go
Comment (randhid) [inline components/base/wheeled/kinematics.go:25]: [q] Does this function need to be public anymore?

File: components/base/wheeled/kinematics.go
Comment (randhid) [inline components/base/wheeled/kinematics.go:27]: [minor] Does
```suggestion
	wb, ok := base.(*wheeledBase)
```
work? in preparation for UnWrap proxy being removed, since we're in the same package.

File: components/base/agilex/limo_base.go
Comment (randhid) [inline components/base/agilex/limo_base.go:56]: [q] is there any way we can keep these unexported? Or will that make it impossible to know which frames are attached to each base in fake?

**Review by randhid — CHANGES_REQUESTED** [review]

File: services/slam/slam.go
Comment (raybjork) [reply services/slam/slam.go]: Per our discussion this is not a useful thing to add to the slam package.  It has been removed

Comment (raybjork) [reply services/slam/slam.go]: Thanks for summarizing our convo!  Removed the function from the slam package

File: components/base/agilex/limo_base.go
Comment (raybjork) [reply components/base/agilex/limo_base.go:57]: This is the pattern we settled on for arms, so I figured it would make sense to apply it to bases too.  Without exporting these we would be relying on hardcoding these strings in the fake package, which is a lot worse IMO.  What is your hesitation with exporting?

File: components/base/wheeled/kinematics.go
Comment (raybjork) [reply components/base/wheeled/kinematics.go:27]: This does not work as the type is a `reconfigurableBase`

File: components/base/wheeled/kinematics.go
Comment (raybjork) [reply components/base/wheeled/kinematics.go:25]: Yes, as it will be used by the motionplan package in a subsequent PR that relies on this work

File: components/base/wheeled/kinematics.go
Comment (randhid) [reply components/base/wheeled/kinematics.go:27]: Wait, is it? reconfigurables are in the `base.go` package. Actually, isn't this passed in from line 23?

Comment (raybjork) [reply components/base/wheeled/kinematics.go:27]: Actually good catch.  This was leftover code from before

File: components/base/agilex/limo_base.go
Comment (randhid) [reply components/base/agilex/limo_base.go:57]: Just that there are two patterns going on - one to unexport as much as we can from each package because that keeps things more under control for someone else importing these packages for their own code. It's not a blocker, just something to think about when deciding what to import/export.

File: components/base/fake/fake.go
Comment (raybjork) [reply components/base/fake/fake.go:31]: So hard code it as just being a wheeledBase always, and delete the Config struct entirely? I'm good with this just wanted to check this is what you are suggesting

File: components/base/agilex/limo_base.go
Comment (raybjork) [reply components/base/agilex/limo_base.go:57]: This could go away for non-wheeled bases if you are good with hardcoding fake as wheeled for now

File: components/base/fake/fake.go
Comment (biotinker) [inline components/base/fake/fake.go:101]: Would be neat if this was configurable.

Even if not configurable, should be a `const` up top rather than a magic number in the middle of the file.

File: components/base/fake/fake.go
Comment (randhid) [reply components/base/fake/fake.go:31]: yep

File: components/base/agilex/limo_base.go
Comment (randhid) [reply components/base/agilex/limo_base.go:57]: I am okay with this for now. Just be aware this might need to be re-tooled in the future, but we'll cross that bridge if/when we get there.

File: components/base/fake/fake.go
Comment (raybjork) [reply components/base/fake/fake.go:101]: Yeah. I copy pasted this from the original file.  I will make this less bad

**Review by nicksanford — APPROVED** [review]
approving as this PR no longer changes slam components to unblock

File: components/base/fake/fake.go
Comment (raybjork) [reply components/base/fake/fake.go]: Made it a constant.  Theres a larger project here that involves making this fully customizable but since Rand requested I hold off on some of those changes I'm going to do the simple thing here

File: components/base/wheeled/kinematics.go
Comment (biotinker) [inline components/base/wheeled/kinematics.go:63]: This feels weird. Shouldn't this live with `Fake` then?

File: components/base/wheeled/kinematics.go
Comment (raybjork) [reply components/base/wheeled/kinematics.go]: I think this is a product of cartographer but I'm not entirely sure.

Comment (biotinker) [reply components/base/wheeled/kinematics.go]: What are min/max Y on these maps? Both zero?

Since we mostly are only using cartographer at the moment, I'm fine with leaving this generally, but would like if we add an error if either +/- Z limits are both 0, or if any Y limit is *not* zero

File: components/base/base.go
Comment (raybjork) [reply components/base/base.go:306]: Actually yeah the more I think about it the more I like this

File: components/base/wheeled/kinematics.go
Comment (randhid) [reply components/base/wheeled/kinematics.go]: 2D (carto) will keep Y the same along the map it has built, and XZ forms the plane of the map. This can obviously change with 3D slam. 

I don't think it is generally enforced. @nicksanford? Long therm there needs to be a transform sent in from the lidar and/or algorithm. Or this is just the contract that we formalize, but that seems tricky to impose.  Maybe a comment/todo is appropriate here.

File: components/base/base.go
Comment (raybjork) [reply components/base/base.go:306]: Actually this code is pretty specific to bases at least at the moment so I am going to pass on this suggestion for the time being.  I will keep this in the back of my head going forward though as we will surely be iterating on this portion

File: components/base/wheeled/kinematics.go
Comment (nicksanford) [reply components/base/wheeled/kinematics.go]: My understanding (which may well be incomplete) is that right now there is no contract specifying which dimensions non-viam-carto 2d slam will use.

File: components/base/wheeled/wheeled_base.go
Comment (randhid) [inline components/base/wheeled/wheeled_base.go:342]: [nit] would rather the variable spelled out in the comments, i know f2 probably renamed them in vscod.

**Review by randhid — APPROVED** [review]
LGTM

**Pull request merged**
